applications/Chat/coati/dataset/sft_dataset.py

```diff
@@ -71,6 +71,38 @@ def _preprocess(sources: Sequence[str],
     return sequences_token["input_ids"], labels, sequences_token["attention_mask"]
 
 
+def _preprocess_chatglm(sources: Sequence[str],
+                targets: Sequence[str],
+                tokenizer: PreTrainedTokenizer,
+                max_length: int,
+                ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    """
+    Preprocess the data by tokenizing.
+    None for attention mask, ChatGLM will calculate attention mask according to input ids
+    """
+  
+    labels = []
+    input_ids = []
+    for source, target in zip(sources, targets):
+        source_id = tokenizer.encode(text=source, add_special_tokens=False)
+        target_id = tokenizer.encode(text=target, add_special_tokens=False)
+        if len(source_id) + len(target_id) >= max_length - 2:
+            source_id = source_id[ : max(0, max_length - len(target_id) - 2)]
+        if len(target_id) > max_length - len(source_id) - 2:
+            target_id = target_id[: max(0, max_length - len(source_id) - 3)] + [target_id[-1]]
```

Comment:
A bit confused about this part. 

It can be sample, you can first build the sequence and then remove the extra from the left.

Reply:
Fixed